services/multilang-support/src/utils.py

```diff
@@ -0,0 +1,218 @@
+
+# Depending on the model chosen from HuggingFace,
+# the tokenizer and model will be different.
+# See README.md#Implementation for more details.
+MODEL_CHECKPOINT = "Helsinki-NLP/opus-mt-en-fr"
+
+
+# Set device to GPU if available, else CPU
+# Issue: torch can detect cuda GPU but will not be able to use it
+# if the GPU is out of memory. (Happened during unicorn testing)
+
+
+@log
+def set_device(device_id: int = 0):
+    """
+    Set the device to GPU if available, else CPU.
+    For GPUs: device_id starts from 0 and increments.
+    For CPUs: device_id is -1.
+
+    To handle the issue of torch detecting cuda GPU but not able to use it,
+    we set the device to try out all GPUs before falling back to CPU.
+    """
+    global DEVICE, DEVICE_NAME
+    if device_id >= 0:
+        DEVICE = torch.device(f"cuda:{device_id}")
+        DEVICE_NAME = torch.cuda.get_device_name(device=DEVICE)
+    else:
+        DEVICE = torch.device("cpu")
+        DEVICE_NAME = "CPU"
+
+    LOGGER.debug(f"Using {DEVICE} on {DEVICE_NAME}")
+    return True
+
+
+# Tokenizer and Model ready to be instantiated
+@log
+def instantiate():
+    global TOKENIZER, MODEL
+    global DEVICE, DEVICE_NAME
+    num_gpus = torch.cuda.device_count()
+    LOGGER.info(f"Instantiating: {MODEL_CHECKPOINT} tokenizer and model")
+
+    TOKENIZER = AutoTokenizer.from_pretrained(MODEL_CHECKPOINT)
+    LOGGER.debug("Tokenizer instantiated")
+
+    MODEL = AutoModelForSeq2SeqLM.from_pretrained(MODEL_CHECKPOINT)
+    LOGGER.debug("Model instantiated")
+
+    device_id = 0
+    while device_id < num_gpus:
+        try:
+            set_device(device_id=device_id)
+            device_id += 1
+            MODEL = MODEL.to(DEVICE)
+            # if it is able to set model to device, it means that GPU is usable
+            # otherwise we will try the next GPU
+            break
+        except Exception as e:
+            LOGGER.warning(f'Error using {DEVICE_NAME}: {e}')
+            if device_id >= num_gpus:
+                LOGGER.warning("No GPU available, using CPU.")
+                set_device(device_id=-1)
+                break
+    LOGGER.debug(f"Model is running on {DEVICE_NAME}.")
+
+
+@log
+def tokenize_query_to_tensor(query: str):
+    """
+    STEP 1: Tokenize the query using the tokenizer.
+    @param query: The query (type<str>) to be tokenized.
+    @return: The tokenized query as a torch.Tensor.
+    """
+    # return TOKENIZER.encode(query).to(DEVICE)
+    return TOKENIZER(query, return_tensors="pt").to(DEVICE).input_ids
+
+
+@log
+def generate_output_tensor(input_ids: torch.Tensor) -> torch.Tensor:
+    """
+    STEP 2: Translate the input_ids tensor to an output query.
+    @param input_ids: The decoded tensor (type<torch.Tensor>) to be translated.
+    @return: Newly generated tensor using the model (type<torch.Tensor>).
+    TODO: Add parameters to control/optimize the translation.
+    """
+    # Parameters:
+    MAX_TIME = 0.5  # seconds
```

Comment:
Recommend not mixing in constants like this throughout the file.

Reply:
Noted.
I originally didn't want to hardcode a constant like this, it will sometimes ruin the output result. However, without parameters like `max_time` or `max_new_tokens`, it produces a warning (below). Having large `max_new_tokens` might slows down the generation time, while low `max_time` prevents long query to be translated.
```yaml
UserWarning: Using `max_length`'s default (512) to control the generation length. This behaviour is deprecated and will be removed from the config in v5 of Transformers -- we recommend using `max_new_tokens` to control the maximum length of the generation.
```
I suggest creating a work item (issue) on what to do with this, will discuss with @rbazin on a solution and maybe adapt JAX implementation instead of current one.

Reply:
To clarify, I'm not opposed to constants! Just I prefer to have them all listed in a few places near the top of a file so that we can quickly make changes in the future. Digging through to find them is unpleasant, especially if you're not the one who wrote the code in the first place!

Reply:
Moved the constant to the top of the file. If there is new change or update to functionality, future PRs will cover them.